mara_pipelines/pipelines.py

```diff
@@ -81,24 +81,49 @@ def html_doc_items(self) -> List[Tuple[str, str]]:
 
 
 class Task(Node):
-    def __init__(self, id: str, description: str, commands: Optional[List[Command]] = None, max_retries: Optional[int] = None) -> None:
+    def __init__(self, id: str, description: str, commands: Optional[Union[Callable, List[Command]]] = None, max_retries: Optional[int] = None) -> None:
         super().__init__(id, description)
-        self.commands = []
+        self.callable_commands = callable(commands)
         self.max_retries = max_retries
 
-        for command in commands or []:
-            self.add_command(command)
-
-    def add_command(self, command: Command, prepend=False):
+        if self.callable_commands:
+            self._commands = None
+            self.__commands_callable = commands
+        else:
+            self._commands = []
+            self._add_commands(commands or [])
+
+    def _test_is_not_dynamic(self):
+        if self.callable_commands:
+            raise Exception('You cannot use add_command when the task is constructed with a callable commands function.')
+
+    @property
+    def commands(self) -> List:
+        if not self._commands:
+            self._commands = []
+            # execute the callable command function and cache the result
+            for command in self.__commands_callable() or []:
```

Comment:
I'm trying to figure out if this caching is save or not. I'm basically unsure if this method could be called before it is actually executed, e.g.` commands` is called which looks into a table to figure out the number of tasks to run in parallel, a load task loads stuff which adds more to the table, the task is run but now misses some data because the commands are cached.

Reply:
I see your point. This is an interesting question because it raises other questions:
1. should it be allowed to call the function multiple times? (this is what I wanted to prevent)
2. should the self.commands be turned into an enumerable instead of a list?
3. what would a parallel task look like?
4. should we implement a logic which prevents calling the function outside of "self.run()"? Meaning outside of a pipeline execution.

I would say for now, that a task in itself should run just a fixed, freezed set of commands. I just was building this quickly so my thoughts didn't go longer than that.

When one want to be more generic, it is maybe time to think about a dynamic pipeline (not yet implemented) or to use a ParallelTask (see #74 as well).

What do you think @jankatins ? Do you have a specific use case in mind?
Sounds like you plan to do some streaming-like activity with mara :-) 

p.s. I build #74 to do some complex semi-streaming file based activity. It works fine now, but had to put a lot of work into it.

Reply:
I've actually never had a usercase where I needed a dynamic list (part from reads, but these were already covered).

Personally I would leave out caching in the first iteration and/or would add a log when it is executed. Then lets see what would actually execute it.

Reply:
OK. I leave it for now with caching since I see it more dangerous to call the function twice.